Add language guide and entry in table for Dart (#4667)
circleci/circleci-docs: jekyll/_cci2/language-dart.md

---
layout: classic-docs
title: "Language Guide: Dart"
short-title: "Dart"
description: "Overview and sample config for a Dart project"
categories: [language-guides]
order: 2
---

> This document is a walkthrough of a sample Dart project setup on CircleCI. The project cherry-picks `number_guesser` and `number_thinker` code from [Write HTTP clients & servers](https://dart.dev/tutorials/server/httpserver) tutorial.

Reviewer: ```suggestion
This document is a walkthrough of a sample Dart project setup on CircleCI. The project cherry-picks `number_guesser` and `number_thinker` code from the [Write HTTP clients & servers](https://dart.dev/tutorials/server/httpserver) tutorial.
```

We assume you have a basic working knowledge of CircleCI and associated terminology. If not, visit our [Getting Started docs](http://0.0.0.0:4000/docs/2.0/getting-started/#section=getting-started).

## Quickstart

> **Repository**: <https://github.com/CircleCI-Public/circleci-dart-demo>   

Reviewer: Can you make these into markdown formatted links?

**Builds**: <https://app.circleci.com/pipelines/github/CircleCI-Public/circleci-dart-demo>

1. Find the full configuration at the bottom of this document or in the repository linked above.
1. Copy that config into [`.circleci/config.yml`]({{ site.baseurl }}/2.0/configuration-reference/) in your project's root directory.
1. Modify as needed.

The CircleCI config does the following:

> - **It runs the tests on a Docker container.**

Reviewer: ```suggestion
- **It runs the tests in a Docker container.**
```

>     - The `test` job uses the `google/dart` Docker image as its execution image (CircleCI doesn't currently have a native Dart image).

Reviewer: ```suggestion
    - The `test` job uses the `google/dart` Docker image as its execution image.
```
I think it's unnecessary to add that we do not have an image.

>     - The tests uses the [junitreporter](https://pub.dev/packages/junitreport) package to produce JUnit XML output for CircleCI's [test metadata feature](https://circleci.com/docs/2.0/collect-test-data/), which in turn supports things like test summary, intelligent test splitting, and insights data/metrics.

Reviewer: ```suggestion
    - The tests use the [junitreporter](https://pub.dev/packages/junitreport) package to produce JUnit XML output for CircleCI's [test metadata feature](https://circleci.com/docs/2.0/collect-test-data/), which in turn supports features such as test summary, intelligent test splitting, and insights data/metrics.
```

    - `~/.pub-cache` and `.dart_tool` folders are cached by default on Linux and macOS. `~/AppData/Local/Pub/Cache` if Windows.
    - For Dart projects that have it, you'll probably also want to add and cache the `.packages` folder in the main project directory.

If you fork this project and want to push to DockerHub, this project assumes [a context](https://circleci.com/docs/2.0/contexts/) called `dart-docker` with the following variables & keys:

| KEY           | VALUE                             |
|---------------|-----------------------------------|
| DOCKER_TAG    | The tag/repository for your image |
| DOCKER_LOGIN  | Your Docker login                 |
| DOCKER_PWD    | Your Docker password              |
{: class="table table-striped"}

See the config and modify as needed for your use case.

## Config Walkthrough

The first section of the file defines common items. The order in which things are defined in a CircleCI configuration doesn't matter - we have laid it out this way for readability.

```yaml
version: 2.1

#####################
# Common Definitions
#####################

# Orb declarations
orbs:
  win: circleci/windows@2.4.0
```

> - **Config Version**: Current is 2.1. Unless using legacy server install or legacy jobs and workflows, we highly suggest using 2.1.

Reviewer: ```suggestion
- **Config Version**: Current is `2.1`. Unless using a legacy server installation,  jobs or workflows, we recommend using 2.1.
```

    parameters:
      shell:
        type: string
        default: "/bin/bash --login -eo pipefail"
    steps:
      - run:
          name: Native compile
          shell: << parameters.shell >>
          command: dart2native bin/server.dart -o circleci_dart_demo.exe
      - store_artifacts:
          path: circleci_dart_demo.exe
```

{% endraw %}

- CircleCI offers [reusable configuration syntax]({{ site.baseurl }}/2.0/reusing-config), which is what our orbs are built off of. Under `commands`, we define a few parameterized commands that run the same sequence of steps.
    - `dependencies` checks out the code and restores/saves a dependency cache (Dart packages downloaded from last build).
    - `native-build` uses the `dart2native` command to compile a native executable for each operating system and then artifacts it, making it available to download from the job details.
    - In both commands, we take parameters to override a cache path and shell for different jobs. Dart's package cache is in a different location in Windows, and it also uses a different shell (Powershell).

```yaml

> ######################

Reviewer: I think I would recommend removing the large comment blocks - the sections are already broken up, and none of our language guides use them.

# Workflow Definition
######################
workflows:
  version: 2.1
  test-and-build:
    jobs:
      - test
      - build-mac:
          requires:
            - test
      - build-windows:
          requires:
            - test
      - build-linux:
          requires:
            - test
      - build-docker:
          # Uncomment/modify context to use secrets (e.g., DockerHub credentials, etc.)
          #context: dart-docker
          requires:
            - test
```

- In the `workflows` section, we define a workflow called `test-and-build` and define which jobs are run.
- Each of the `build-*` jobs depend on `test` passing, which is set by the `requires` key.

> - There's a commented line for `build-docker`, which is using a `dart-docker` context. See links to resources at the bottom of this document.

Reviewer: ```suggestion
- The commented line for `build-docker` uses a `dart-docker` context. See links to resources at the bottom of this document.
```
